packages/languages/src/common/language-server-service.ts

```diff
@@ -0,0 +1,40 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+// tslint:disable:no-any
+
+/**
+ * The WS endpoint path to the LanguageServer service.
+ */
+export const LanguageServerPath = '/services/language-server';
+
+/**
+ * LanguageServerService symbol for DI.
+ */
+export const LanguageServerService = Symbol('LanguageServerService');
+
+/**
+ * This service provides functionality to set parameters before starting a new language server process.
+ */
+export interface LanguageServerService {
+    /**
+     * Sets parameters before starting ls.
+     *
+     * @param contributionPath path to the language contribution
+     * @param parameters parameters which the client sends to the server before the start request
+     */
+    setStartParameters(contributionPath: string, parameters: any[]): Promise<void>;
```

Comment:
Each contribution has an ID which should match on the frontend and backend. How about using it instead of `path`?

Reply:
What if several pages in parallel setting parameters and starting language servers? How identity between pages and parameters is preserved?

I wonder can we instead introduce something like:
```
/** create a new language server session. */
create(parameters: any): Promise<string>;
/** destroy a language server session by id */
destroy(id: string): Promise<void>;
```

This id then should be passed as a part of the contribution path: `/languages/java/100`.
The language server contribution will receive an id and retrieve parameters based on it.

Reply:
@svor I've reimplemented it with https://github.com/theia-ide/theia/pull/2970#discussion_r234454908 approach here: https://github.com/theia-ide/theia/pull/3538 With it, there should not be races between multiple pages.

Each server contribution gets the start options as a second argument:
https://github.com/theia-ide/theia/pull/3538/commits/667aa1bbf228391fc35cfc76224261077873b4ed#diff-0d434f8608475b2bff20baa430b67637

Each client contribution should implement `startParameters`, this getter is called on each start: https://github.com/theia-ide/theia/pull/3538/commits/667aa1bbf228391fc35cfc76224261077873b4ed#diff-f5eb2e9d4635ab8f2ac59fc2f1e2c626R192

Reply:
@akosyakov You are right it could be a problem with multiple pages in my changes. Thank you for providing the solution. So what should I do with this PR? Is it better to close this PR and create another which will provide startParameters for Java LS or is it better to rework current one when https://github.com/theia-ide/theia/pull/3538 will be merged?

Reply:
I hope to merge #3538 today/tomorrow. You can rebase this PR.